packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1459 @@
+    /**
+     * @notice Emitted when KRO undelegation is initiated.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param amount    The amount of KRO to undelegate.
+     * @param shares    The amount of shares to undelegate.
+     */
+    event KroUndelegationInitiated(
+        address indexed validator,
+        address indexed delegator,
+        uint128 amount,
+        uint128 shares
+    );
+
+    /**
+     * @notice Emitted when KGH undelegation is initiated.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param tokenId   Token Id of the KGH.
+     * @param kroShares The amount of KRO shares to undelegate.
+     * @param kghShares The amount of KGH shares to undelegate.
+     */
+    event KghUndelegationInitiated(
+        address indexed validator,
+        address indexed delegator,
+        uint256 tokenId,
+        uint128 kroShares,
+        uint128 kghShares
+    );
+
+    /**
+     * @notice Emitted when KGHs undelegation is initiated in batch.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param tokenIds  Array of token Ids of the KGHs.
+     * @param kroShares The amount of KRO shares to undelegate.
+     * @param kghShares The amount of KGH shares to undelegate.
+     */
+    event KghBatchUndelegationInitiated(
+        address indexed validator,
+        address indexed delegator,
+        uint256[] tokenIds,
+        uint128 kroShares,
+        uint128 kghShares
+    );
+
+    /**
+     * @notice Emitted when KRO undelegation is finalized.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param amount    The amount of KRO undelegated.
+     */
+    event KroUndelegationFinalized(
+        address indexed validator,
+        address indexed delegator,
+        uint128 amount
+    );
+
+    /**
+     * @notice Emitted when KGH undelegation is finalized.
+     *
+     * @param validator Address of the validator.
+     * @param delegator Address of the delegator.
+     * @param amount    The amount of KRO undelegated during the KGH undelegation.
+     */
+    event KghUndelegationFinalized(
+        address indexed validator,
+        address indexed delegator,
+        uint128 amount
+    );
+
+    /**
+     * @notice Emitted when pending challenge reward for challenge winner is distributed.
+     *
+     * @param recipient Address of the reward recipient.
+     * @param amount    The amount of challenge reward.
+     */
+    event ChallengeRewardDistributed(address recipient, uint128 amount);
+
+    /**
+     * @notice Emitted when the output reward is distributed.
+     *
+     * @param recipient Address of the reward recipient.
+     * @param amount    The amount of reward.
+     */
+    event RewardDistributed(address recipient, uint128 amount);
```

Comment:
Why don't we indexing the recipient address?

Reply:
I missed that point, will add one. Thanks!